### ISSUE TITLE
What happens when multiple people are using the meditor on the same dandiset simulataneously?

### ISSUE THREAD

**Issue by mvandenburgh** [dandi/dandiarchive#831]
We may want to consider some kind of locking mechanism to prevent people trampling each other's edits.

**Comment by satra** [comment]
we may also want to consider something like https://togetherjs.com/ for owners who are logged in and click the metadata editor.

**Comment by yarikoptic** [comment]
note that it might be not "people" but "us (some metadata migration/mutation/enrichment directly within dandi-api) + a user" so ideally it might need to be at dandi-api level

**Comment by waxlamp** [comment]
I thought the easiest thing to do would be to implement a simple lock on the metadata that would need to be obtained before a client could edit the metadata, but the problem of stale locks or of computers crashing while holding the lock, etc., seem rather difficult to solve properly.

Perhaps the next simplest thing to do is to hold a copy of the metadata as it was at the time the meditor is opened, then retrieve a fresh copy when the user wants to save. If the saved and new copied differ, warn the user of the change, show them the "diff", and then return them to edit mode. They can then click "save" again to write their version of metadata (or, in general, continue editing to account for the changes they see). This is better UX than bluntly using locks to ensure only one editor at a time, but it might take some design work to get it right.

This is one of those things that we need to solve, but maybe don't need to solve _right now_. Nonetheless, discussing some solution designs now might be a good idea.

**Comment by yarikoptic** [comment]
re locks: I would expect that there is some "heartbeat" solution in web tech, i.e. lock is released unless heartbeat from the client is not received within e.g. 1 minute, and client sends them e.g. every 30 seconds and expects to receive a reconfirmation that lock is still "its".

Re "simplest" one -- I believe it is similar to what github does for e.g. editing a commit message. It would just announce that message was changed while editing.  I think it indeed might be the simplest one.  Showing a diff of edits done by the client would indeed be a nice gesture.  Might be worth even showing the diff of externally done changes to compare.

If we had "tracking" of actions done by users, might be nice to just say "last changed by X on DATETIME" or "last opened for changes by X on DATETIME" (if that is what the log says, i.e. was opened for changes but they were never saved).

**Comment by mvandenburgh** [comment]
> Perhaps the next simplest thing to do is to hold a copy of the metadata as it was at the time the meditor is opened, then retrieve a fresh copy when the user wants to save. If the saved and new copied differ, warn the user of the change, show them the "diff", and then return them to edit mode. They can then click "save" again to write their version of metadata (or, in general, continue editing to account for the changes they see). This is better UX than bluntly using locks to ensure only one editor at a time, but it might take some design work to get it right.

This solution still leaves a similar problem open - what if the metadata changes again after the diff is displayed but before the user clicks "save" again? I suppose we could just show another diff (maybe that is what you were implying already)

**Comment by waxlamp** [comment]
> > Perhaps the next simplest thing to do is to hold a copy of the metadata as it was at the time the meditor is opened, then retrieve a fresh copy when the user wants to save. If the saved and new copied differ, warn the user of the change, show them the "diff", and then return them to edit mode. They can then click "save" again to write their version of metadata (or, in general, continue editing to account for the changes they see). This is better UX than bluntly using locks to ensure only one editor at a time, but it might take some design work to get it right.
> 
> This solution still leaves a similar problem open - what if the metadata changes again after the diff is displayed but before the user clicks "save" again? I suppose we could just show another diff (maybe that is what you were implying already)

Yeah, it would be like a while loop that could theoretically continue forever, but in practice will not (you would need an unreasonable number of owners, all editing at the same time, or a trolling process that's continuously updating the metadata as fast as possible).

**Comment by waxlamp** [comment]
> re locks: I would expect that there is some "heartbeat" solution in web tech, i.e. lock is released unless heartbeat from the client is not received within e.g. 1 minute, and client sends them e.g. every 30 seconds and expects to receive a reconfirmation that lock is still "its".
> 

Interesting idea. I think this would work but it also illustrates why locking is troublesome in general.

> Re "simplest" one -- I believe it is similar to what github does for e.g. editing a commit message. It would just announce that message was changed while editing. I think it indeed might be the simplest one. Showing a diff of edits done by the client would indeed be a nice gesture. Might be worth even showing the diff of externally done changes to compare.
> 

I think in typical cases this would be enough. In general, at most one person will be editing metadata. If it happens that two owners are independently trying to write to the metadata, I think an announcement of this sort would be enough in the vast majority of cases. We should take reasonable steps to make sure that work isn't lost, but that could be accomplished fairly easily just by stashing the user's edits in local storage, etc. (I don't think this needs to be done right away, but it seems like a good idea to implement eventually.)

> If we had "tracking" of actions done by users, might be nice to just say "last changed by X on DATETIME" or "last opened for changes by X on DATETIME" (if that is what the log says, i.e. was opened for changes but they were never saved).

I like this idea. Let's keep it in mind.
